nginx: build offline documentation

KAction opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
nginx: build offline documentation

7c6f434c commented:
Hmm, given the constraints I am not sure if this is better off as a separate `ngnix-doc` package (I guess datestamp-versioned)…

KAction commented:
Can't tell for all users, but I normally expect that documentation is either in $out or in $doc. If it is in neither, I assume it is non-existent.

7c6f434c commented:
(Looks at `man-pages` that are, among other things, doc for `glibc`)

Maybe it would be cleaner to define separate `nginx-doc` and pass it to `nginx` as parameter then? Something with an explicit goal to evaluate twice, separately, to the same result, in some system setups looks a bit weird…

Commit:
nginx-doc-unstable: init at 2022-05-05

Build documentation of nginx (which is maintained separately from webserver
itself) and make it available both as "nginx-doc" attribute and as "doc" output
of "nginx" derivation.

KAction commented:
Okay, I stand corrected, I also check `man` output. man-pages is special in many ways, in particular such that everybody knows that they exist.

Anyway, factored `nginx-doc` into separate top-level attribute, so now documentation is both in `nixpkgs.nginx-doc` and `nixpkgs.nginx.doc`. Can't install both, since they conflict.

7c6f434c commented:
I think if you symlink instead of copying you can actually install both, no?

KAction commented:
Having $doc as symlink by itself does not work: `nix-env -i ./result-doc` succeeds, but ~/.nix-profile/share/doc/nginx does not appear.

KAction commented:
For what it worth, I can set `priority = 6` to `nginx-doc` (just did it).

SuperSandro2000 commented on pkgs/servers/http/nginx/generic.nix:
```diff
@@ -114,8 +117,12 @@ stdenv.mkDerivation {
+  # which breaks build, since nginx does not actually use autoconf.
+  preConfigure = ''
+    setOutputFlags=
+    '' + preConfigure
+       + concatMapStringsSep "\n" (mod: mod.preConfigure or "") modules;
```
```suggestion
  '' + preConfigure
     + concatMapStringsSep "\n" (mod: mod.preConfigure or "") modules;
```

SuperSandro2000 commented on pkgs/data/documentation/nginx-doc/default.nix:
```diff
@@ -0,0 +1,40 @@
+
+  meta = with lib; {
+    description = "A reverse proxy and lightweight webserver (documentation)";
+    homepage    = "http://nginx.org";
```
```suggestion
    homepage    = "https://nginx.org/";
```

7c6f434c commented:
Right, maybe it only works with lndir (when all directories are real, and the files are just symlinks)

KAction commented:
I don't think you can get raw `builtin:buildenv` to ignore conflicts ever:

```
; nix-env -i ./result-doc ./result-2/
replacing old 'nginx-1.20.2'
replacing old 'nginx-doc-unstable-2022-05-05'
installing 'nginx-1.20.2-doc'
installing 'nginx-doc-unstable-2022-05-05'
building '/nix/store/9vgqcx323kry3igxdpl5v4p1si2hdcda-user-environment.drv'...
error: packages '/nix/store/jbh9lcyn99cb6aiql4cyb0x3g5gvml83-nginx-doc-unstable-2022-05-05/share/doc/nginx/en/donation.html' and '/nix/store/6wnb0dyd96yfcwgz59wfjiwvgawwffq7-nginx-1.20.2-doc/share/doc/nginx/en/donation.html' have the same priority 5; use 'nix-env --set-fla
g priority NUMBER INSTALLED_PKGNAME' to change the priority of one of the conflicting packages (0 being the highest priority)
error: packages '/nix/store/jbh9lcyn99cb6aiql4cyb0x3g5gvml83-nginx-doc-unstable-2022-05-05/share/doc/nginx/en/donation.html' and '/nix/store/6wnb0dyd96yfcwgz59wfjiwvgawwffq7-nginx-1.20.2-doc/share/doc/nginx/en/donation.html' have the same priority 5; use 'nix-env --set-fla
g priority NUMBER INSTALLED_PKGNAME' to change the priority of one of the conflicting packages (0 being the highest priority)
builder for '/nix/store/9vgqcx323kry3igxdpl5v4p1si2hdcda-user-environment.drv' failed with exit code 1
error: build of '/nix/store/9vgqcx323kry3igxdpl5v4p1si2hdcda-user-environment.drv' failed
; file /nix/store/6wnb0dyd96yfcwgz59wfjiwvgawwffq7-nginx-1.20.2-doc/share/doc/nginx/en/donation.html
/nix/store/6wnb0dyd96yfcwgz59wfjiwvgawwffq7-nginx-1.20.2-doc/share/doc/nginx/en/donation.html: symbolic link to /nix/store/jbh9lcyn99cb6aiql4cyb0x3g5gvml83-nginx-doc-unstable-2022-05-05/share/doc/nginx/en/donation.html
; file /nix/store/jbh9lcyn99cb6aiql4cyb0x3g5gvml83-nginx-doc-unstable-2022-05-05/share/doc/nginx/en/donation.html
/nix/store/jbh9lcyn99cb6aiql4cyb0x3g5gvml83-nginx-doc-unstable-2022-05-05/share/doc/nginx/en/donation.html: HTML document, Unicode text, UTF-8 text, with very long lines (2499)
```

KAction commented:
So while `nginx-doc` is nice to have top-level for redefining in overlay, it is still mostly implementation detail of `nginx.doc`.

KAction commented:
@7c6f434c @SuperSandro2000 Do I miss anything before we can merge?

SuperSandro2000 commented on pkgs/servers/http/nginx/generic.nix:
```diff
@@ -145,6 +152,12 @@ stdenv.mkDerivation {
 
   enableParallelBuilding = true;
 
+  nginx_doc = nginx-doc;
+  preInstall = ''
+    mkdir -p $doc
+    cp -r $nginx_doc/* $doc
```
```suggestion
  preInstall = ''
    mkdir -p $doc
    ln -s ${nginx_doc}/* $doc
```

KAction replied:
I don't like it. I interacts poorly with `overrideAttrs`.

KAction replied:
Fine.

SuperSandro2000 replied:
You can easily change the doc by overriding nginx-doc. The current code is not that nice.

KAction replied:
It is a bit trickier due double `callPackage`, but okay, I can do it with overrides.

SuperSandro2000 commented on pkgs/data/documentation/nginx-doc/default.nix:
```diff
@@ -0,0 +1,40 @@
+  version = "2022-05-05";
+  src = fetchhg {
+    url = "https://hg.nginx.org/nginx.org";
+    sha256 = "029n4mnmjw94h01qalmjgf1c2h3h7wm798xv5knk3padxiy4m28b";
+    rev = "a3aee2697d4e";
```
```suggestion
    rev = "a3aee2697d4e";
    sha256 = "029n4mnmjw94h01qalmjgf1c2h3h7wm798xv5knk3padxiy4m28b";
```
normally we always try to write the sha last.
